File: 0260-single-number-iii/0260-single-number-iii.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumber = function(nums) {
    const n = nums.length;
    let _xor = 0;
    for(let i = 0; i<n; i++) _xor ^= nums[i];

    let rightBit = _xor & (-_xor);
    let x = 0, y = 0;

    for(let i = 0; i<n; i++) {
        if((rightBit & nums[i]) == 0) x ^= nums[i];
        else y ^= nums[i];
    }

    return [x, y];
};

File: 0260-single-number-iii/0260-single-number-iii.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int n = nums.size();
        long long _xor = 0;
        for(int i = 0; i<n; i++) _xor ^= nums[i];

        long long rightBit = _xor & (-_xor);
        int x = 0, y = 0;

        for(int i = 0; i<n; i++) {
            if((rightBit & nums[i]) == 0) x ^= nums[i];
            else y ^= nums[i];
        }

        return {x, y};
    }
};